### PR TITLE
Import features and mappings to knowledge base

### PR DESCRIPTION
**Pull request by mstajdohar**
```
wget https://data.genialis.com/repo/3f92c79f-888b-4655-9eda-af46a5bd9c1f/b24507989acb9d618ac9a391bb29dfd345daf71a/?file_name=GRCh38.p7_gene_info.tab.zip&op=download
./manage.py insert_features GRCh38.p7_gene_info.tab.zip

wget https://data.genialis.com/repo/3f92c79f-888b-4655-9eda-af46a5bd9c1f/a714a1aeb35b508499dbbdc11f1aeb5126fa5205/?file_name=MGI_to_ncbi_relation_table.tab.gz&op=download
./manage.py insert_mappings MGI_to_ncbi_relation_table.tab.gz

./manage.py rebuild_index
```

### CODE REVIEW

File: resolwe_bio/tests/commands/test_import_kb.py [hunk 1]
```diff
@@ -0,0 +1,27 @@
...
+        call_command('insert_features', os.path.join(TEST_FILES_DIR, 'features_update.tab.gz'))
+        mock_logger.info.assert_called_with('Total features: 4. Inserted 1, updated 1, unchanged 2, failed 0.')
+
+    @mock.patch('resolwe_bio.management.commands.insert_mappings.logger')
+    def test_insert_mappings(self, mock_logger):
+        call_command('insert_mappings', os.path.join(TEST_FILES_DIR, 'mappings.tab'))
+        mock_logger.info.assert_called_with('Total mappings: 5. Inserted 5, unchanged 0.')
+
+        call_command('insert_mappings', os.path.join(TEST_FILES_DIR, 'mappings_update.tab'))
+        mock_logger.info.assert_called_with('Total mappings: 6. Inserted 6, unchanged 0.')
```
Comment (mstajdohar) [inline resolwe_bio/tests/commands/test_import_kb.py]: This should be changed when unique constraint is supported in mappings relation.

Comment (kostko) [inline resolwe_bio/tests/commands/test_import_kb.py]: Unique constraint has been added in genialis/resolwe-bio#168.

File: resolwe_bio/management/commands/insert_mappings.py [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+
+                mapping = Mapping(relation_type=row['relation_type'],
+                                  source_db=row['source_db'],
+                                  source_id=row['source_id'],
+                                  target_db=row['target_db'],
+                                  target_id=row['target_id'])
+                try:
+                    mapping.save()
+                    count_inserted += 1
+                except IntegrityError as exc:
```
Comment (kostko) [inline resolwe_bio/management/commands/insert_mappings.py]: You should use [`update_or_create`](https://docs.djangoproject.com/en/1.9/ref/models/querysets/#update-or-create) instead.

File: resolwe_bio/management/commands/utils.py [hunk 3]
```diff
@@ -150,3 +153,50 @@ def generate_reads_descriptor(organism_name, presample=True):
...
+    elif ext == '.gz':
+        _open = gzip.open
+
+    if _open:
+        with _open(file_name) as tsv_file:
+            line_count = sum(1 for row in tsv_file)
+
+        with _open(file_name) as tsv_file:
+            yield (os.path.basename(file_name), line_count, tsv_file)
+
```
Comment (kostko) [inline resolwe_bio/management/commands/utils.py]: Remove extra spaces in if clauses.

File: resolwe_bio/management/commands/insert_mappings.py (hunk 2 above)
Comment (mstajdohar) [inline resolwe_bio/management/commands/insert_mappings.py]: But can you detect unchanged with this?

File: resolwe_bio/management/commands/utils.py (hunk 3 above)
Comment (mstajdohar) [inline resolwe_bio/management/commands/utils.py]: Done.

File: resolwe_bio/tests/commands/test_import_kb.py (hunk 1 above)
Comment (mstajdohar) [inline resolwe_bio/tests/commands/test_import_kb.py]: Changed.

**Comment by mstajdohar** [comment]
@kostko, your comments have been addressed. Please check again and approve if the PR is now good to merge.

File: resolwe_bio/management/commands/insert_mappings.py [hunk 4]
```diff
@@ -0,0 +1,56 @@
...
+        count_inserted, count_unchanged = 0, 0
+
+        for tab_file_name, line_count, tab_file in decompress(options['file_name']):
+            logger.info(__("Importing mappings from \"{}\":", tab_file_name))
+
+            reader = csv.DictReader(tab_file, delimiter=str('\t'))
+            bar_format = '{desc}{percentage:3.0f}%|{bar}| {n_fmt}/{total_fmt} [{elapsed}<{remaining}]'
+
+            for row in tqdm(reader, total=line_count, bar_format=bar_format):
+
```
Comment (kostko) [inline resolwe_bio/management/commands/insert_mappings.py]: What's with the blank line?

Comment (mstajdohar) [inline resolwe_bio/management/commands/insert_mappings.py]: Fixed, thanks.

**Pull request merged**